### PR TITLE
Check compatibility when autostarting, not before autostart

### PR DESCRIPTION
**Pull request by ravicious**
Responding to this comment: https://github.com/gravitational/teleport/pull/32477#discussion_r1338217821

The check for `!isAgentConfigured` was was added here: https://github.com/gravitational/teleport/pull/31951#discussion_r1331447325

I originally wanted to make it so that if the agent is not started and there's a compatibility error, we show an error indicator. However, the indicator shouldn't be shown if the agent is not configured. Otherwise we'd show the error if versions are incompatible even if the user never touched Connect My Computer.

Instead of adding `isAgentConfigured` inside `case 'not-started'`, I assumed that there's no reason to show the indicator at all if the agent is not configured, hence this guard clause.

What we really want to ask is "Was the agent not autostarted because of a compatibility error?". But the check for compatibility is done from outside of any `useAsync` call, so there's no `attempt` object we could inspect.

I was going to say that this isn't necessarily bad and that if `startAgent` entered an error state because of compatibility issues, we'd have to add a special check for that as well. But is it actually true?

Wouldn't it make the code less complex if `shouldAutoStartAgent` didn't look at agent compatibility, merely check if the compatibility is known, and then `startAgent` would return an error if the versions were incompatible?

---

This PR aims to implement this.

### CODE REVIEW

**Comment by ravicious** [comment]
I tested this using packaged versions with my Cloud cluster by manipulating `-c.extraMetadata.version=` passed to `yarn package-term`.

File: web/packages/teleterm/src/ui/ConnectMyComputer/CompatibilityPromise.tsx [hunk 1]
```diff
@@ -49,7 +49,7 @@ export function checkAgentCompatibility(
     : 'incompatible';
 }
 
-export function CompatibilityError(): JSX.Element {
+export function CompatibilityError(props: { showAlert: boolean }): JSX.Element {
```
Comment (gzdunek) [inline web/packages/teleterm/src/ui/ConnectMyComputer/CompatibilityPromise.tsx:52]: `hideAlert?: boolean` maybe? Usually we want to show the alert, so perhaps it is better to crate a special case for hiding it?

File: web/packages/teleterm/src/ui/ConnectMyComputer/connectMyComputerContext.tsx [hunk 2]
```diff
@@ -355,17 +368,32 @@ export const ConnectMyComputerContextProvider: FC<{
...
+        try {
+          await downloadAndStartAgent();
+        } catch (error) {
+          // Turn off autostart if it fails, otherwise the user wouldn't be able to turn it off by
+          // themselves.
+          workspacesService.setConnectMyComputerAutoStart(
+            rootClusterUri,
+            false
+          );
+        }
```
Comment (gzdunek) [inline web/packages/teleterm/src/ui/ConnectMyComputer/connectMyComputerContext.tsx:395]: 👍

**Comment by gzdunek** [comment]
image

Should we change "Failed to verify agent binary" to "Failed to check agent version"? For the `processing` state we have "Checking agent version", so it would match it more.

**Review by gzdunek — COMMENTED** [review]
It is much better now, thanks!

**Review by gzdunek — APPROVED** [review]

File: web/packages/teleterm/src/ui/ConnectMyComputer/CompatibilityPromise.tsx (hunk 1 above)
Comment (ravicious) [reply web/packages/teleterm/src/ui/ConnectMyComputer/CompatibilityPromise.tsx:52]: Yeah, I think I'm going to do `hideAlert` with a default to false.

**Comment by ravicious** [comment]
> Should we change "Failed to verify agent binary" to "Failed to check agent version"? For the `processing` state we have "Checking agent version", so it would match it more.
> 
> EDIT: Or maybe not, it could be confusing.

Right, the biggest hurdle is that it's not only checking the version but also downloading it.

**Pull request merged**